### ISSUE TITLE
Add <pgFootLast>

### ISSUE THREAD

**Issue by pe-ro** [music-encoding/music-encoding#572]
Currently, MEI can deal with a page header and footer on the first page, and on the following pages.  However, sometimes a page footer on the last page is unique; that is, different from those on the preceding pages.  Adding `<pgFootLast>` would make it possible to represent this situation.

For example:
```
<scoreDef>
  <pgHead>Top of first page</pgHead>
  <pgHead2>Top of subsequent pages</pgHead2>
  <pgFoot>Foot of first page</pgFoot>
  <pgFoot2>Foot of subsequent pages</pgFoot2>
  <pgFootLast>Foot of last page</pgFootLast>
</scoreDef>
```
could be used for a score with a unique header on the first and last pages and repeating headers and footers on the intervening pages.  In this case, `<pgFoot2>` should be replaced by `<pgFootLast>` on the last page.

The following example captures unique header on page 1 and unique footer on the last page:
```
<scoreDef>
  <pgHead>Top of first page</pgHead>
  <pgFootLast>Foot of last page</pgFootLast>
</scoreDef>
```
`<pgHeadLast>` could also be added, but I haven't yet seen an real-life examples of this.

**Comment by ahankinson** [comment]
`<pgFoot type="last">`? Or something along those lines?

**Comment by pe-ro** [comment]
Nope.  That assigns special meaning to a particular value of `@type`, which is a no-no.

What's your objection to `<pgFootLast>`?

**Comment by ahankinson** [comment]
The "something along those lines" meant "maybe not type, but something else?"

I tend to think of elements as "things" and attributes as "and this what makes this thing different than the others." It's still a page footer -- it's just a specific type of page footer.

**Comment by pe-ro** [comment]
That works -- sometimes.

But at what point does a "thing of a certain type" become a new thing?  Horse-drawn carts and jet airplanes are both types of conveyance, but are vastly different in practical terms.  In this case, a pattern has already been established -- we have `<pgHead>` for a header on the first page, `<pgHead2>` for headers on subsequent pages, `<pgFoot>` for footers on the first page, and `<pgFoot2>` for footers on pages other than the first.  In other words, different elements for different "things".  Adding a new `<pgFootLast>` seems appropriate to me.  This relieves processors from the burden of dealing with a mix of controlled and uncontrolled values of `@type`.

Adding a different attribute ("something along those lines" of `@place`) to a single `<headFoot>` element could work, but that would be pretty drastic now, wouldn't it?  And, for this to work, the new element would have to be repeatable, meaning there could be multiple elements with `@place="head"`.  Again, the processor has to figure out what to do or there has to be additional info in the encoding.  See how this is getting more complex, not simpler?

**Comment by ahankinson** [comment]
I wouldn't necessarily say that 'pgFoot' and 'pgFoot2' are great examples to follow either, but wasn't going to open that can of worms :). If you want to take lessons from page layout, you would have footers for even/odd (left/right in an opening), and then perhaps one for a single page. The first and last are then styled based on being, well, first and last.

Would we also need to have pgHeadLast, pgHeadFirst, and pgFootFirst?

What about just `pgFoot` with `@position = first | last | even | odd`?

**Comment by pe-ro** [comment]
You're in charge now.  Do what you feel is appropriate.

**Comment by craigsapp** [comment]
>  you would have footers for even/odd (left/right in an opening)

That is a good point.

> However, sometimes a page footer on the last page is unique;

What example case of this?  Related to this I am wondering where someone would place lyrics for secondary verses that are sometimes given after the music? Verovio has page footers implemented, so they can go there (if the last footer is available), but that is a hack in terms of what a footer usually means.

And somewhat related, I get requests for footnotes.  Is there a way of encoding them and controlling how they would be inserted into (I presume) the footer for a page?  This would also be related to the last footer, since the foot notes could instead be rendered as endnotes.

**Comment by ahankinson** [comment]
> You're in charge now. Do what you feel is appropriate.

What I think is appropriate is to have a conversation about it and work through it. Don't you think that's the best way to tackle this?

**Comment by t-bullock** [comment]
Any new thoughts on this? It would be nice to see support for different headers or footers on even and odd pages.

**Comment by lpugin** [comment]
I agree it would be nice to have this and I like the proposal to support this with `pgHead` and `pgFoot` with `@position` rather than adding specific elements. It seems to me that it would be logical to drop `pgHead2` and `pgFoot2`.

Any objections?

**Comment by pe-ro** [comment]
What are the proposed values of ***@***.***’?

--
p.

**Comment by lpugin** [comment]
I would suggest `@position` = `first | second | last | even | odd` ?

* `first` would correspond to the current `pgHead` / `pgFoot`
* `second` would correspond to the current `pgHead2` / `pgFoot2`

**Comment by pe-ro** [comment]
OK.  How will duplicate values of ***@***.***’ be handled?  Assuming such things will be illegal, what mechanism will disallow/handle the following –

<pgHead position=”first”>…</pgHead>
<pgHead position=”first”>…</pgHead> ?

Also, “even” and “odd” depend on physical page numbering or assumed/auto-generated page numbering.  This can present problems.  For example, with the following markup --

<pgHead position=”first”>…</pgHead>
<pgHead position=”even”>…</pgHead>
<pgHead position=”odd”>…</pgHead>


and the first page labeled “2”; that is, there is no page labeled “1”, which of the headers takes precedence?  Also, what’s the anticipated result on the so-called odd pages that carry even page numbers/labels and vice versa?

In addition, what’s anticipated in the markup above if <pgHead position=”odd”> isn’t present?

I think that neither the current schema nor the proposal can address headers/footers on first/intervening/last pages and odd/even page headers/footers simultaneously without introducing ambiguities.  Certainly, a complex system of Schematron rules can be devised or conventions can be documented that deal with these situations, but I still think the simplest, most effective solution is an element-based one.  That approach provides opportunities to control order and number that attributes do not.

--
p.

**Comment by lpugin** [comment]
> OK.  How will duplicate values of ***@***.***’ be handled?  Assuming such things will be illegal, what mechanism will disallow/handle the following –
```xml
<pgHead position=”first”>…</pgHead>
<pgHead position=”first”>…</pgHead>
```
> ?

I hoped that this could be controlled via Schematron. But I know very little about it, and you seem to imply that it cannot. That is too bad.

> Also, “even” and “odd” depend on physical page numbering or assumed/auto-generated page numbering.  This can present problems.  For example, with the following markup --
```xml
<pgHead position=”first”>…</pgHead>
<pgHead position=”even”>…</pgHead>
<pgHead position=”odd”>…</pgHead>
```
> and the first page labeled “2”; that is, there is no page labeled “1”, which of the headers takes precedence?  Also, what’s the anticipated result on the so-called odd pages that carry even page numbers/labels and vice versa?

Could you say more about the page labelling markup you have in mind here? I am not aware of strong structuring concept for page labelling in MEI. My understanding is that a page label would be simply some text as it appears on the page, assuming that in reality you can have odd numbers on physical even pages, and reverse. You are right that it can be ambiguous, and maybe we should not have 'odd' and 'even' attribute values, but instead 'left' and 'right'. (We probably then need to have a way to encode that it is left first, or right first. I think this would be a quite common typographic approach.)

I would note that this is not any more problematic than the current solution. It seems to me that we can have the same type of ambiguity, for example with a `<pgHead>` (the running header of the first page) related to a page labelled "2", or `pgHead2` with a page labelled "1". As far as I know, this can currently not be avoided. (And we cannot control that the page labelled "2" comes after page labelled "1", and that is absolutely not a problem in my opinion.)

> In addition, what’s anticipated in the markup above if `<pgHead position=”odd”>` isn’t present?

I remember having encountered a similar issue when implementing `<pgHead>` and `<pgHead2>`. I cannot remember what Verovio does when there is no `<pgHead2>`, but I do remember that I had to take a somehow arbitrary decision on showing `<pgHead>` on the second page, or not showing anything at all (because no `<pgHead2>` is actually given). Both would arguably be valid behaviours. What did you actually anticipated in such a case? 

I agree we might need to clarify some of these issues, but they are essentially guidelines ones, and many are ambiguities that we already have. I do not think they are bound to the fact that the proposed solution is using a single element with an attribute, rather than distinct elements. It seems to me that we can avoid here the multiplication of elements, assuming that this is always something we want to avoid when it is not strictly necessary.

**Comment by pe-ro** [comment]
Since I haven’t tried it, I can’t say that using Schematron wouldn’t be possible, I just don’t think it’s the most efficient approach.  We’d essentially be trading a complex structure (multiple elements) for a different kind of complexity (complex Schematron).  The Schematron would likely be very complex as it would have to enforce number and order of the pared-down set of elements simultaneously.

There isn’t a strong concept of page labeling in MEI, per se.  Explicit page labels are expected to be provided using processing instructions within page headers/footers, for example,

<scoreDef>
  <pgHead><?page-number "3"?></pgHead>
</scoreDef>

in the case of auto-generated numbers or

<pb>
  <pgHead>3</pgHead>
</pb>

on every page in the case of explicitly encoded numbers.

The problem with “odd” and “even” is more of a user confusion issue.  Either of the preceding examples could be a so-called “even” page, leading to confusion about which header/footer should appear on the page.  Using elements and term “alternating”, avoids this issue, I think.  “Left/right” or “recto/verso” would be better values for ***@***.***”, but there’s still the problem of controlling the complexity of elements with various values of ***@***.***” with Schematron.  With the current solution, an encoder could still label the 2nd page with a “3”, but a great deal of confusion can be eliminated by not using the terms “odd” and “even”.  Note that the current definitions of <pgHead>, <pgFoot>, etc. are careful not to use these terms, preferring “first”, “last”, “intervening”, “between”, “alternating”, etc.

To illustrate my thoughts about expected behaviors, I’ve attached a simplified markup file with annotations covering various cases.  Let me know if it doesn’t make through to github.

Clear, succinct markup employing as few elements as possible is a great organizing principle.  However, principles must be bent sometimes for practical reasons.  I think this is one of those cases.

**Comment by lpugin** [comment]
The example did not make it through. You can paste it here if it is now too long.

I believe that, if Schematron can constrain what we want, then we should go for it, even if it is complex. It has to be done once, and it would be a shame to go for a less good modelling solution just because of that.

I still feel that `pgHeadXXX` is not quite right. It does really look like concatenating an attribute to the element name, and since XML has attributes, it would better be put in one of them. That seems to be an even more fundamental principle we should follow. You would not do `<staff1>`, `<staff2>` as a replacement of `<staff n="1">` and `<staff n="2">` just because that would allow you to control their order (assuming you would want that) without Schematron, would you?

**Comment by pe-ro** [comment]
Here’s the example markup –

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE test SYSTEM "file:/C:/Users/pdr4h/OneDrive%20-%20University%20of%20Virginia/Desktop/pgHeadFootTest.dtd">
<test>

  
  <scoreDef/>

  
 <scoreDef>
    <pgHead><?page-number "3"?></pgHead>
  </scoreDef>

  
  <scoreDef>
    <pgHead/>
    <pgFoot/>
  </scoreDef>

  
  <scoreDef>
    <pgHead/>
    <pgFoot2/>
  </scoreDef>

  
  <scoreDef>
    <pgHead2/>
    <pgFoot2/>
  </scoreDef>

  
  <scoreDef>
    <pgHead2/>
    <pgFootLast/>
  </scoreDef>

  
  <scoreDef>
    <pgHead/>
    <pgHead2/>
  </scoreDef>

  
  <scoreDef>
    <pgFoot/>
    <pgFoot2/>
  </scoreDef>

  
  <scoreDef>
    <pgHead/>
    <pgHead/>
    <pgHead2/>
  </scoreDef>

  
  <scoreDef>
    <pgHead/>
    <pgHead2/>
    <pgHeadLast/>
  </scoreDef>

  
  <scoreDef>
    <pgHead/>
    <pgHead/>
    <pgHead2/>
    <pgHeadLast/>
  </scoreDef>

  
  <scoreDef>
    <pgHead/>
    <pgHead/>
    <pgFoot/>
    <pgHead2/>
    <pgFoot2/>
    <pgHeadLast/>
  </scoreDef>

  
  <scoreDef>
    <pgHead/>
    <pgHeadLast/>
  </scoreDef>

  
  <scoreDef>
    <pgHeadLast/>
    <pgFootLast/>
  </scoreDef>

</test>

The important part of the schema behind the markup is –

<!ELEMENT scoreDef (pgHead?, pgHead?, pgFoot?, (pgHead2?, pgFoot2?)?, pgHeadLast?, pgFootLast?) >

You’re right, I wouldn’t use <staff1>, <staff2>, etc., but this is a more complex case.  In the case of <staff>, order can be established by @n.  There’s more at play here than simpler order, though.

--
p.

**Comment by pe-ro** [comment]
Another possibility is to use a generic element, “fw” perhaps and allow the current element names as values in the position attribute; that is –

position := head | foot | head2 | foot2| headLast | footLast

and use Schematron to control number of occurrences of each; that is, 2 occurrences of head, 1 each of the others.  It would be left to the processor to figure out the meaning of the various combinations of position values.

A variation of this is to use pgHead and pgFoot elements with position values of “first | last | alt1 | alt2”.  I think I like this more, but it still puts an interpretive burden on the processor.

**Comment by lpugin** [comment]
> A variation of this is to use pgHead and pgFoot elements with position values of “first | last | alt1 | alt2”.  I think I like this more, but it still puts an interpretive burden on the processor.

Sounds good. I assume `alt` means `alternative`. Correct? Or `alternate`?  Or do you want to keep it deliberately ambiguous?

**Comment by pe-ro** [comment]
alt = alternating

The value “alt1” and “alt2” indicate alternating (left/right, verso/recto, odd/even) form work, but without reliance on page numbering/labeling. The presence of “alt1” and “alt2” indicates an alternating sequence of alt1, alt2, alt1, alt2, etc.  If only “alt1” is present, then the patten is assumed to be alt1, [blank], alt1, [blank], etc.  Similarly, if only “alt2” is present, then the pattern is [blank], alt2, [blank], alt2, etc.  “first” and “last” interrupt this pattern.  If “first” is present, then the alternating pattern begins on the next page following the first.  If “last” is present, then the last page will have the indicated header/footer despite what the next one in the pattern would be.  If only “first” and “last” are present, then the alternating pattern never takes place.

Make sense?

**Comment by lpugin** [comment]
It does make sense. I think that for backward compatibility we need a replacement for what `pgHead2` currently does. That is, defining a header from the second page onwards, but without alternating. If I understand correctly, one would need to define both `alt1` and `alt2` (i.e., without gap). That is probably not ideal. Any thoughts?

**Comment by pe-ro** [comment]
Yeah, this needs more thought


_________________________________________
Perry Roland, MA, MLIS
Metadata Operations Librarian
University of Virginia Library
2450 Old Ivy Rd.
Charlottesville, VA 22903
434-982-2702 (w)
pdr4h (at) virginia (dot) edu

* Office phone forwarded to cell
________________________________

**Comment by pe-ro** [comment]
The best thing I can come up with at the moment is to add another value, such as “afterFirst”, “followingFirst”, or “all” -- where “all” actually means “all the pages except the first”.

I’d also suggest using “place” rather than “position” as the name of the attribute that controls where the header / footer should be displayed.  <pgHead> and <pgFoot> should NOT be members of the existing att.placement class, however.  Either the place attribute should be declared locally; that is, within the definition of both elements, or a new attribute class should be created with a definition of the place attribute that allows the proper values for these elements.

**Comment by pe-ro** [comment]
The best thing I can come up with at the moment is to add another value, such as “afterFirst”, “followingFirst”, or “all” -- where “all” actually means “all the pages except the first”.

**Comment by lpugin** [comment]
I would actually now leave it as it is. In terms of backward compatibility, if someone has data where only `<pgHead2>` is encoded - which as I understand is the only problematic case - we can recommend to add an empty header for the first page in order to encode that `all` applies actually only from the second page onward:
```xml
<pgHead func="first"/>
<pgHead func="all">
   <!-- The content of pgHead2 in MEI 4.0 -->
</pgHead>
```
Does it work? @musicEnfanthen ?

**Comment by musicEnfanthen** [comment]
Since, according to the description, `first` interrupts the pattern of `all`, this seems to exclude the empty first pgHead

**Comment by lpugin** [comment]
Excatly, this is the idea. Works for you?

**Comment by musicEnfanthen** [comment]
Yes, works for me in general. (However, I am wondering if there are other cases where we use an empty element to indicate that something does not exist? But maybe that is more a philosophical issue...)

**Comment by lpugin** [comment]
Yes, you are right that this is probably not perfect. But keep in mind that this is only for backward compatibility and for cases that are quite likely pretty rare.

**Comment by pe-ro** [comment]
The only other case for using an empty element to indicate something missing that I can think of now is the case where one source has something another doesn't.  For a markup example, see the Guidelines, 11.1.2 Variants in Musical Content.

**Comment by bwbohl** [comment]
@pe-ro good example; and for the more philosophical thoughts: `<pubStmt/>`(which is required but may be empty)

**Comment by musicEnfanthen** [comment]
Closed by #1008
